ramips: mt76x8: add support for Cudy LT400E v1

kjonca opened the pull request:
This commit adds support for the Cudy LT400E v1,
 an indoor 4G/LTE WiFi router.
It has 4x10/100 Mbps Ethernet ports, a 2.4 GHz radio, and a Quectel 4G/LTE EC200A modem.

Hardware:

    SoC: MediaTek MT7628NN
    Flash: 8 MiB SPI
    RAM: 64 MiB
    Ethernet: 4x10/100 Mbps
    WiFi: 2.4 GHz (embedded MT7628, 802.11n)
    WWAN: Quectel EC200A-EL 4G/LTE Cat 4 modem
    SIM: nano-SIM slot
    Buttons: 1 Reset button, 1 WPS button
    LEDs: 10x (blue)
    Power: 12 V DC barrel jack

Installation:
The stock firmware is a customized variant of OpenWrt, which implements a signature check that only allows flashing official firmware. Cudy offers, on their website [1][6], intermediate OpenWrt firmware images which do not implement the signature check. After flashing the intermediate image, the upstream (i.e., official) OpenWrt image can be installed.  See the device's entry in OpenWrt's Wiki for more details [2].

Cellular network:
The Quectel EC200A modem is internally connected to the SoC via USB, and operates in ECM mode, supported by the CDC Ether driver [3]. When inserting a SIM card in the slot, if no PIN is needed, the network interface "usb0" is brought up and the cellular connection works out of the box. To interact with the modem directly (unlock the PIN, change APN and PDP settings, etc.), AT commands can be sent with picocom to serial port /dev/ttyUSB2 [4].

Recovery:
The stock firmware can be recovered via TFTP, as the bootloader (based on U-Boot) implements a recovery client. Set up a TFTP server on your computer with IP 192.168.88.1/24 serving the stock firmware from Cudy's website renamed to "recovery.bin". Press the Reset button (keep it pressed), power on the device, wait for the TFTP server to send the recovery.bin file, then release the Reset button. The router will take a couple of minutes to reboot and set up the stock firmware. See the device's entry in OpenWrt's Wiki for more details [2].

Discussion:
Use the dedicated thread in OpenWrt's forum to discuss any aspects [5].

Acknowledgments:
The code to support this device is largely based on the sources kindly provided by Cudy in compliance with the GPL license. Thanks for supporting the OpenWrt community.

Links:
[1] https://www.cudy.com/blogs/faq/openwrt-software-download
[2] https://openwrt.org/toh/cudy/lt400e_v1
[3] https://openwrt.org/docs/guide-user/network/wan/wwan/ethernetoverusb_cdc
[4] https://openwrt.org/docs/guide-user/network/wan/wwan/at_commands
[5] https://forum.openwrt.org/t/242466
[6] https://drive.google.com/drive/folders/1BKVarlwlNxf7uJUtRhuMGUqeCa5KpMnj

hauke commented on target/linux/ramips/dts/mt7628an_cudy_lt400e-v1.dts:
```diff
@@ -0,0 +1,223 @@
+		compatible = "gpio-leds";
+
+		led_power: power {
+			label = "blue:power";
```
I think this would match it better:
```suggestion
			function = LED_FUNCTION_POWER;
			color = <LED_COLOR_ID_BLUE>;
```

hauke commented on target/linux/ramips/image/mt76x8.mk:
```diff
@@ -199,6 +199,17 @@ define Device/creality_wb-01
+  DEVICE_MODEL := LT400E
+  DEVICE_VARIANT := v1
+  DEVICE_PACKAGES := kmod-usb2 kmod-usb-ohci kmod-usb-net-cdc-ether \
+  kmod-usb-serial-option picocom
```
Why do you need `picocom`?  Please remove it from the default options if it is not needed for basic operation.

kjonca commented on target/linux/ramips/image/mt76x8.mk:
```diff
@@ -199,6 +199,17 @@ define Device/creality_wb-01
+  DEVICE_VARIANT := v1
+  DEVICE_PACKAGES := kmod-usb2 kmod-usb-ohci kmod-usb-net-cdc-ether \
+  kmod-usb-serial-option picocom
+  SUPPORTED_DEVICES += cudy,lt400e
```
Honestly - I don't know, where can I check it?

kjonca replied:
> Amending `SUPPORTED_DEVICES` is meant to allow compatibility with OpenWrt-based vendor firmware. Please try if the vendor intermediate firmware accepts an image created _without_ amending `SUPPORTED_DEVICES` and if so, please drop this line.

So I should:
1. compile image without this line 
2. install vendor's image
3. try to install my compiled openwrt?

Commit:
ramips: mt76x8: add support for Cudy LT400E v1

This commit adds support for the Cudy LT400E v1,
an indoor 4G/LTE WiFi router.
It has 4x10/100 Mbps Ethernet ports, a 2.4 GHz radio,
and a Quectel 4G/LTE EC200A modem.

Hardware:

SoC: MediaTek MT7628NN
Flash: 8 MiB SPI
RAM: 64 MiB
Ethernet: 4x10/100 Mbps
WiFi: 2.4 GHz (embedded MT7628, 802.11n)
WWAN: Quectel EC200A-EL 4G/LTE Cat 4 modem
SIM: nano-SIM slot
Buttons: 1 Reset button, 1 WPS button
LEDs: 10x (blue)
Power: 12 V DC barrel jack

Installation:
The stock firmware is a customized variant of OpenWrt, which implements
a signature check that only allows flashing official firmware. Cudy
offers, on their website [1][6], intermediate OpenWrt firmware images
which do not implement the signature check. After flashing the
intermediate image, the upstream (i.e., official) OpenWrt image can be
installed.  See the device's entry in OpenWrt's Wiki for more details [2].

Cellular network:
The Quectel EC200A modem is internally connected to the SoC via USB, and
operates in ECM mode, supported by the CDC Ether driver [3]. When
inserting a SIM card in the slot, if no PIN is needed, the network
interface "usb0" is brought up and the cellular connection works out of
the box. To interact with the modem directly (unlock the PIN, change APN
and PDP settings, etc.), AT commands can be sent with picocom to serial
port /dev/ttyUSB2 [4].

Recovery:
The stock firmware can be recovered via TFTP, as the bootloader (based
on U-Boot) implements a recovery client. Set up a TFTP server on your
computer with IP 192.168.88.1/24 serving the stock firmware from Cudy's
website renamed to "recovery.bin". Press the Reset button (keep it
pressed), power on the device, wait for the TFTP server to send the
recovery.bin file, then release the Reset button. The router will take a
couple of minutes to reboot and set up the stock firmware. See the
device's entry in OpenWrt's Wiki for more details [2].

Discussion:
Use the dedicated thread in OpenWrt's forum to discuss any aspects [5].

Acknowledgments:
The code to support this device is largely based on the sources kindly
provided by Cudy in compliance with the GPL license. Thanks for
supporting the OpenWrt community.

Links:
[1] https://www.cudy.com/blogs/faq/openwrt-software-download
[2] https://openwrt.org/toh/cudy/lt400e_v1
[3] https://openwrt.org/docs/guide-user/network/wan/wwan/ethernetoverusb_cdc
[4] https://openwrt.org/docs/guide-user/network/wan/wwan/at_commands
[5] https://forum.openwrt.org/t/242466
[6] https://drive.google.com/drive/folders/1BKVarlwlNxf7uJUtRhuMGUqeCa5KpMnj

Co-authored-by: rogerpueyo <roger.pueyo@guifi.net>

Signed-off-by: Kamil Jońca <kjonca@onet.pl>
Link: https://github.com/openwrt/openwrt/pull/21246
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

carlosz1986 commented:
@kjonca thx for your work :-)
I have started to add support for the [lt700 lte poe version.](https://github.com/openwrt/openwrt/pull/24689)
The "Cellular Signal Strength" leds are not triggered yet and wanted to ask you if you found a solution for your device?
Otherwise I will implement something proper and wanted to ask you if you are up for testing it later..
cheers :-)
